### PR TITLE
Change EGM code to be able to read HCAL PF thresholds from DB

### PR DESCRIPTION
**Pull request by RSalvatico**
#### PR description:

This is a PR with changes in the EGM code to allow HCAL PF cuts to be taken from DB on demand (i.e., for specific workflows) in alternative to being taken from config files (as discussed [here](https://indico.cern.ch/event/1220970/contributions/5284605/attachments/2598482/4486203/HCAL_PFlowCutsSeeds_Feb24_2023.pdf)).

Lots of adaptation work was done already on the PF side and in several aspects this should be similar.
https://github.com/cms-sw/cmssw/pull/43025

#### PR validation

Several Run3 and Run2 workflows (e.g. runTheMatrix.py -l 10024.0, runTheMatrix.py -l 10824.0...)

#### Expected changes

None, modulo that the thresholds provided via GT are the same as previously provided via config file (which should be the case).

### CODE REVIEW

**Comment by missirol** [comment]
@RSalvatico

Since "EGM code" could possibly include also code used by EGM at HLT, could you please let us know if the following plugins used at HLT should also be updated to use `HcalPFCuts` ? (If so, will this be done in a future PR ?)

```
EgammaHLTHcalVarProducerFromRecHit
FixedGridRhoProducerFastjetFromRecHit
```

FYI: @cms-sw/hlt-l2

**Comment by swagata87** [comment]
> FixedGridRhoProducerFastjetFromRecHit

I'm taking care of this.

**Comment by swagata87** [comment]
@missirol https://github.com/cms-sw/cmssw/pull/43176 here is the recHit rho code migration

File: RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc [hunk 1]
```diff
@@ -471,6 +488,20 @@ GEDPhotonProducer::GEDPhotonProducer(const edm::ParameterSet& config, const Cach
...
 }
 
+void GEDPhotonProducer::beginRun(const edm::Run& run, const edm::EventSetup& es){
+  if (cutsFromDB) {
+    const HcalTopology& htopo = es.getData(htopoToken_);
+    const HcalPFCuts& hcalCuts = es.getData(hcalCutsToken_);
+
+    std::unique_ptr<HcalPFCuts> paramPF_;
+    paramPF_ = std::make_unique<HcalPFCuts>(hcalCuts);
+    paramPF_->setTopo(&htopo);
```
Comment (mmusich) [inline RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:498]: See https://github.com/cms-sw/cmssw/pull/43025#discussion_r1381646747

**Comment by RSalvatico** [comment]
> @RSalvatico
> 
> Since "EGM code" could possibly include also code used by EGM at HLT, could you please let us know if the following plugins used at HLT should also be updated to use `HcalPFCuts` ? (If so, will this be done in a future PR ?)
> 
> ```
> EgammaHLTHcalVarProducerFromRecHit
> FixedGridRhoProducerFastjetFromRecHit
> ```
> 
> FYI: @cms-sw/hlt-l2

@missirol I am trying to take care of the other one, EgammaHLTHcalVarProducerFromRecHit, in this PR.

File: RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc (hunk 1 above)
Comment (RSalvatico) [reply RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:499]: Yes, I'll address this

File: RecoEgamma/EgammaIsolationAlgos/python/egammaHBHERecHitThreshold_cff.py [hunk 2]
```diff
@@ -7,11 +7,20 @@
...
+#    recHitEThresholdHB = _thresholdsHBphase1_2023
+#)
+
+# from Configuration.Eras.Modifier_run3_egamma_2023_cff import run3_egamma_2023
+# run3_egamma_2023.toReplaceWith(egammaHBHERecHit,egammaHBHERecHit_2023)
+
+from Configuration.Eras.Modifier_run3_egamma_cff import run3_egamma
+from Configuration.Eras.Modifier_phase2_common_cff import phase2_common
+from Configuration.Eras.Modifier_run3_hcal_cff import run3_hcal
+run3_hcal.toModify(egammaHBHERecHit,
```
Comment (mmusich) [inline RecoEgamma/EgammaIsolationAlgos/python/egammaHBHERecHitThreshold_cff.py:23]: In https://github.com/cms-sw/cmssw/pull/43025 a different modifier was used `hcalPfCutsFromDB`. I think it should be used uniformly across areas.

**Comment by RSalvatico** [comment]
please test with #43139

**Comment by RSalvatico** [comment]
type egamma

**Comment by RSalvatico** [comment]
please test with #43139

**Comment by RSalvatico** [comment]
test parameters

- pull_request = https://github.com/cms-sw/cmssw/pull/43095

**Comment by RSalvatico** [comment]
please test

File: RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc [hunk 3]
```diff
@@ -118,18 +124,28 @@ class HcalChannelPropertiesEP : public edm::ESProducer {
     return prod;
   }
 
+  ReturnType3 produce3(const HcalPFCutsRcd& rcd) {
```
Comment (mmusich) [inline RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc:127]: this is changed also at https://github.com/cms-sw/cmssw/pull/43025. Why duplicating it here?
Would it make sense to de-scope the technical changes to the RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc in another technical PR or wait for #43025 to be merged?

Comment (RSalvatico) [reply RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc:127]: Sure, I can wait for #43025 to be merged. I needed this to test my changes though, at least locally (afaik there is no other way).

**Comment by RSalvatico** [comment]
> @RSalvatico
> 
> I am curious about the command you gave at [#43164 (comment)](https://github.com/cms-sw/cmssw/pull/43164#issuecomment-1810337176). PR #43095 is already merged, so I don't think adding that changes in any way the results. I believe the failures are coming from the code introduced in this PR.

@mmusich you're evidently right. I wasn't sure whether that was needed or not, so I tried. It looks like all the RelVals failing are 2017, so I thought perhaps there was something wrong with conditions I wasn't picking up. I'm trying to debug...

File: RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc (hunk 3 above)
Comment (mmusich) [reply RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc:127]: > (afaik there is no other way).

please test with https://github.com/cms-sw/cmssw/pull/43025 should work ;)

Comment (RSalvatico) [reply RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc:127]: Yes but I think runTheMatrix.py -l # will not work, that's what I meant with locally

Comment (mmusich) [reply RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc:127]: > Yes but I think runTheMatrix.py -l # will not work, that's what I meant with locally

you can pull in other commits, but avoid pushing them to the central fork when duplicated.

**Comment by RSalvatico** [comment]
test parameters

**Comment by RSalvatico** [comment]
please test

**Comment by antoniovilela** [comment]
@RSalvatico @cms-sw/hlt-l2 @cms-sw/reconstruction-l2 
Do you think it would be reasonable for this PR to converge by tomorrow (Fri)? This is not meant to push the review, just to understand if we wait for it in 13_3_0. 
Thanks.

**Comment by mmusich** [comment]
> Do you think it would be reasonable for this PR to converge by tomorrow (Fri)? 

This PR duplicates code from https://github.com/cms-sw/cmssw/pull/43025 that I think takes precedence, so I think that's unlikely.

**Comment by RSalvatico** [comment]
@antoniovilela @mmusich I also don't mean to push the review, but how is the duplication of code the bottleneck? I left the duplicated code here because it wasn't clear to me anymore whether the PR #43025 was going to be merged by today/tomorrow, in which case this duplicated code would be needed. If you say this is the reason why it can't converge by tomorrow, I'll remove all the duplicates straight away.

**Comment by mmusich** [comment]
> also don't mean to push the review, but how is the duplication of code the bottleneck? 

Either one of the two needs to rebase once the other is merged, and honestly I haven't even started to look at this one more in depth, so I would not reverse the order of merging.

> was going to be merged by today/tomorrow

I think #43025 is almost ready, I plan to sign for HLT after the next round of test pass. 

> why it can't converge by tomorrow

That's not the only reason, but I would like to understand the urgency.

**Comment by perrotta** [comment]
Suggestion: put the duplicate code in a separate extra PR, and test this together with both that extra PR and https://github.com/cms-sw/cmssw/pull/43025. Then if https://github.com/cms-sw/cmssw/pull/43025 converges in time, these two can be merged together, otherwise this and the new extra PR can go first (and https://github.com/cms-sw/cmssw/pull/43025 will need to be rebased)

**Comment by mmusich** [comment]
> put the duplicate code in a separate extra PR, 

Isn't this what I suggested at https://github.com/cms-sw/cmssw/pull/43164#discussion_r1393842414 ?

File: Configuration/Eras/python/Era_Run3_cff.py [hunk 4]
```diff
@@ -15,7 +15,8 @@
 from Configuration.Eras.Modifier_run2_HLTconditions_2018_cff import run2_HLTconditions_2018
 from Configuration.Eras.Modifier_run3_RPC_cff import run3_RPC
 from Configuration.Eras.Modifier_run3_ecal_cff import run3_ecal
+from Configuration.Eras.Modifier_hcalPfCutsFromDB_cff import hcalPfCutsFromDB
```
Comment (mmusich) [inline Configuration/Eras/python/Era_Run3_cff.py:18]: this is also duplicated from https://github.com/cms-sw/cmssw/pull/43025

File: RecoEgamma/EgammaElectronAlgos/interface/ElectronHcalHelper.h [hunk 5]
```diff
@@ -36,17 +38,19 @@ class ElectronHcalHelper {
     int maxSeverityHB;
     EgammaHcalIsolation::arrayHE eThresHE;
     int maxSeverityHE;
+    //const HcalPFCuts* hcalCuts;
```
Comment (mmusich) [inline RecoEgamma/EgammaElectronAlgos/interface/ElectronHcalHelper.h:41]: needed?

File: RecoEgamma/EgammaElectronAlgos/interface/ElectronHcalHelper.h [hunk 6]
```diff
@@ -36,17 +38,19 @@ class ElectronHcalHelper {
...
 
   void beginEvent(const edm::Event &evt, const edm::EventSetup &eventSetup);
 
   inline auto hcalTowersBehindClusters(const reco::SuperCluster &sc) const { return egamma::towersOf(sc, *towerMap_); }
-  double hcalESum(const reco::SuperCluster &, int depth) const;
+  double hcalESum(const reco::SuperCluster &, int depth, const HcalPFCuts *hcalCuts) const;
   double hOverEConeSize() const { return cfg_.hOverEConeSize; }
   int maxSeverityHB() const { return cfg_.maxSeverityHB; }
   int maxSeverityHE() const { return cfg_.maxSeverityHE; }
+  //const HcalPFCuts *hcalCuts() const { return cfg_.hcalCuts; }
```
Comment (mmusich) [inline RecoEgamma/EgammaElectronAlgos/interface/ElectronHcalHelper.h:53]: needed?

File: RecoEgamma/EgammaElectronAlgos/src/ElectronHcalHelper.cc [hunk 7]
```diff
@@ -34,6 +34,7 @@ void ElectronHcalHelper::beginEvent(const edm::Event& evt, const edm::EventSetup
                                                      0.,
                                                      cfg_.eThresHB,
                                                      EgammaHcalIsolation::arrayHB{{0., 0., 0., 0.}},
+                                                     //cfg_.hcalCuts,
```
Comment (mmusich) [inline RecoEgamma/EgammaElectronAlgos/src/ElectronHcalHelper.cc:37]: needed?

File: RecoEgamma/EgammaElectronAlgos/src/ElectronHcalHelper.cc [hunk 8]
```diff
@@ -51,6 +52,7 @@ void ElectronHcalHelper::beginEvent(const edm::Event& evt, const edm::EventSetup
...
```
Comment (mmusich) [inline RecoEgamma/EgammaElectronAlgos/src/ElectronHcalHelper.cc:55]: needed?

File: RecoEgamma/EgammaPhotonProducers/src/PhotonProducer.cc [hunk 9]
```diff
@@ -110,6 +114,11 @@ class PhotonProducer : public edm::stream::EDProducer<> {
   std::unique_ptr<ElectronHcalHelper> hcalHelperCone_;
   std::unique_ptr<ElectronHcalHelper> hcalHelperBc_;
   bool hcalRun2EffDepth_;
+
+  edm::ESGetToken<HcalPFCuts, HcalPFCutsRcd> hcalCutsToken_;
+  bool cutsFromDB;
+  HcalPFCuts* paramPF;
```
Comment (mmusich) [inline RecoEgamma/EgammaPhotonProducers/src/PhotonProducer.cc:120]:
```suggestion
```

unused?

**Review by mmusich — COMMENTED** [review]
from a very diagonal and cursory look

**Comment by RSalvatico** [comment]
@mmusich thanks for the comments, I'll address them and descope this PR from the changes which are common between this and #43025. We were asked by PPD to do our best to make this converge before the closure of 13_3 (not the reviewers' fault if this doesn't happen, needless to say). I guess the idea was that it would be nice to validate it with the 13_3_X validation campaign.

**Comment by mmusich** [comment]
> I guess the idea was that it would be nice to validate it with the 13_3_X validation campaign.

How do I reconcile this with

> No validation has been done yet - still need some work before testing

In the PR description?

**Comment by RSalvatico** [comment]
> > I guess the idea was that it would be nice to validate it with the 13_3_X validation campaign.
> 
> How do I reconcile this with
> 
> > No validation has been done yet - still need some work before testing
> 
> In the PR description?

The PR description stayed the same from the moment I opened a draft PR two weeks ago - since then I've validated it with several workflows. I should have changed it.

**Comment by antoniovilela** [comment]
@RSalvatico @cms-sw/hlt-l2 @cms-sw/reconstruction-l2 
@cms-sw/ppd-l2 @cms-sw/orp-l2 
After consulting with PPD, it seems the best is to descope this PR, and run the Winter24 production with the thresholds from the configuration. Riccardo, could you run it by Jordan and Helena, so that everyone is on the same page?
Thanks.

**Comment by RSalvatico** [comment]
> @RSalvatico @cms-sw/hlt-l2 @cms-sw/reconstruction-l2 @cms-sw/ppd-l2 @cms-sw/orp-l2 After consulting with PPD, it seems the best is to descope this PR, and run the Winter24 production with the thresholds from the configuration. Riccardo, could you run it by Jordan and Helena, so that everyone is on the same page? Thanks.

Yes, confirmed with Jordan and Helena that we'll now target 14_0_0 for this. Thanks.

I'll still make another PR with the common technical changes between this one and the PF one.

**Comment by RSalvatico** [comment]
I should have reverted the technical changes needed by both this PR and #43025:
`Configuration/Eras/python/Era_Run3_cff.py`
`Configuration/Eras/python/Modifier_hcalPfCutsFromDB_cff.py`
`RecoLocalCalo/HcalRecAlgos/plugins/HcalChannelPropertiesEP.cc`

Now this PR requires #43305 to be tested.

**Comment by RSalvatico** [comment]
please test with #43305

**Comment by mmusich** [comment]
@RSalvatico 
there seem to be several changes in offline electron reconstruction (see e.g. [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-16-1100+580732/59807/validateJR.html)). Please clarify if these are expected.

**Comment by swagata87** [comment]
> @RSalvatico there seem to be several changes in offline electron reconstruction (see e.g. [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-16-1100+580732/59807/validateJR.html)). Please clarify if these are expected.

this is indeed strange, and looking at `RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc`, and other codes that make use of it, it seems there was a bug in the old way of applying thresholds, which is uncovered here. The new way of applying thresholds is probably bug-free. I did not thoroughly check things, but I'd like to still write down here my concern that the old way seem to have a bug.

EGamma can do further debugging. One easy way to check is that, if you print the thresholds obtained in old way from here:
https://github.com/cms-sw/cmssw/blob/634657ce95bae7d8866a100ca29952a9a30e1b16/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L151-L152
and the thresholds obtained in new way here:
https://github.com/cms-sw/cmssw/blob/634657ce95bae7d8866a100ca29952a9a30e1b16/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L156-L158
you immediately notice that thresholds are different. Then if you start digging why, it seems the depth info was not correct in old way. I'm not sure if the following two lines are needed:
https://github.com/cms-sw/cmssw/blob/634657ce95bae7d8866a100ca29952a9a30e1b16/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L133
and 
https://github.com/cms-sw/cmssw/blob/634657ce95bae7d8866a100ca29952a9a30e1b16/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L147
And `depth` is provided from other codes in a weird way. The second argument supposed to be depth, but is it so?
https://github.com/cms-sw/cmssw/blob/634657ce95bae7d8866a100ca29952a9a30e1b16/RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc#L364-L366

One has to sit down and understand the details of what's going on here. I did not have time for that right now.

File: RecoEgamma/EgammaIsolationAlgos/python/egammaHBHERecHitThreshold_cff.py [hunk 10]
```diff
@@ -7,11 +7,11 @@
...
-egammaHBHERecHit_2023 = egammaHBHERecHit.clone(
-    recHitEThresholdHB = _thresholdsHBphase1_2023
-)
+from Configuration.Eras.Modifier_hcalPfCutsFromDB_cff import hcalPfCutsFromDB
+hcalPfCutsFromDB.toModify(egammaHBHERecHit,
+                   usePFThresholdsFromDB = True)
+
 
-from Configuration.Eras.Modifier_run3_egamma_2023_cff import run3_egamma_2023
-run3_egamma_2023.toReplaceWith(egammaHBHERecHit,egammaHBHERecHit_2023)
```
Comment (swagata87) [inline RecoEgamma/EgammaIsolationAlgos/python/egammaHBHERecHitThreshold_cff.py:17]: I think it's not a good idea to remove these. If someone wants to run on 2023 samples with `usePFThresholdsFromDB = cms.bool(False)`, then they will get wrong thresholds of HCal, as these lines are deleted. I suggest to keep it for backward compatibility reasons.

**Comment by swagata87** [comment]
> > @RSalvatico there seem to be several changes in offline electron reconstruction (see e.g. [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-16-1100+580732/59807/validateJR.html)). Please clarify if these are expected.
> 
> this is indeed strange, and looking at `RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc`, and other codes that make use of it, it seems there was a bug in the old way of applying thresholds, which is uncovered here. 

I think this is the bug that I try to fix here: https://github.com/cms-sw/cmssw/pull/43355

**Comment by RSalvatico** [comment]
Indeed @swagata87, what you fix in #43355 seems to be the problem. With that, I see no differences in the thresholds retrieved from DB and from config. Also, going back to this:

https://github.com/cms-sw/cmssw/blob/29909891e150c9781f4ade2a6f7b5beb0bd67a6e/RecoEgamma/EgammaIsolationAlgos/src/EgammaHcalIsolation.cc#L132

I think it makes sense because the HCAL depths start from 1, but clearly when reading the thresholds from config one needs to retrieve the first number from an array, hence the 0th entry.

**Comment by RSalvatico** [comment]
please test with #43355

**Comment by mmusich** [comment]
> DQMHistoTests: [Total failures: 1382](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-23-1100+580732/59918/validateJR.html)

are these expected even when running https://github.com/cms-sw/cmssw/pull/43355 ?

**Comment by mmusich** [comment]
test parameters:

**Comment by mmusich** [comment]
@RSalvatico please squash the commits to a minimum compatible with the development history, before the next round of tests.

**Comment by RSalvatico** [comment]
> > DQMHistoTests: [Total failures: 1382](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2023-11-23-1100+580732/59918/validateJR.html)
> 
> are these expected even when running #43355 ?

I'm trying to understand. It's not just that the differences are still there, they have largely increased (and there was almost no change in the code from my side). And I see differences concerning reco tracks that I don't understand: they weren't generated by this PR before and they aren't generated by #43355.
Investigating...

**Comment by RSalvatico** [comment]
Sorry for the noise, I'll fix the squash.

**Comment by RSalvatico** [comment]
Still two files to restore, cannot test yet

**Comment by RSalvatico** [comment]
please test

**Comment by mandrenguyen** [comment]
@RSalvatico Can you please specify what changes are expected in the PR description?

**Comment by RSalvatico** [comment]
> @RSalvatico Can you please specify what changes are expected in the PR description?

@mandrenguyen Sure, done. The answer is that I don't expect changes, if the HCAL PF thresholds provided via GT are the same as those in the old config (which should be the case) and provided that the comparison is made wrt a release where #43355 is integrated (which I thought was the case). Maybe I am wrong on the latter point?

**Comment by perrotta** [comment]
> > @RSalvatico Can you please specify what changes are expected in the PR description?
> 
> @mandrenguyen Sure, done. The answer is that I don't expect changes, if the HCAL PF thresholds provided via GT are the same as those in the old config (which should be the case) and provided that the comparison is made wrt a release where #43355 is integrated (which I thought was the case). Maybe I am wrong on the latter point?

The CMSSW_14_0_X_2023-11-24-1100 IB used for the [tests](https://github.com/cms-sw/cmssw/pull/43164#issuecomment-1826092885) already includd @43355
Therefore I conclude that the chenges observed are **not expected**, and must be understood before merging this PR

File: RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc [hunk 11]
```diff
@@ -500,10 +504,10 @@ GsfElectronAlgo::EventData GsfElectronAlgo::beginEvent(edm::Event const& event,
           egHcalIsoConeSizeOutSmall,
           EgammaHcalIsolation::InclusionRule::withinConeAroundCluster,
           egHcalIsoConeSizeIn,
-          hcalHelperCone_.eThresHB(),
+          EgammaHcalIsolation::arrayHB{{0., 0., 0., 0.}},
```
Comment (RSalvatico) [inline RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc:507]: @mandrenguyen I think this and the following similar changes are the reason why we still see differences. I just meant to keep these lines untouched (they are fixed in #43355), but in the meantime this PR was basically rebased, so by doing this I am modifying them again.

**Comment by mmusich** [comment]
@RSalvatico this PR has again 7 commits (8 taking into account the projected future commit described in https://github.com/cms-sw/cmssw/pull/43164#discussion_r1405878762). Please consider squashing again (1-2 commits max). Thank you.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
+hlt
   * for https://github.com/RSalvatico/cmssw/commit/aee26b236e80ef251f311d73f9dc01c3dc92715d
   * PR according to description and follow-up review;
   * no unexpected changes in PR tests (changes [12434.7](https://cmsweb.cern.ch/dqm/dev/start?runnr=1;dataset%3D/RelVal_wf12434_7_pr/CMSSW_14_0_X-PRcmssw_43164-59954/DQMIO;referenceshow%3Dall;referencenorm=False;referenceobj1%3Dother::/RelVal_wf12434_7_base/CMSSW_14_0_X-PRcmssw_43164-59954/DQMIO::;sampletype%3Doffline_relval;workspace%3DEverything;) are spurious);

File: RecoEgamma/EgammaPhotonProducers/src/ConvertedPhotonProducer.cc [hunk 12]
```diff
@@ -181,6 +193,10 @@ void ConvertedPhotonProducer::beginRun(edm::Run const& r, edm::EventSetup const&
 
   // Transform Track into TransientTrack (needed by the Vertex fitter)
   transientTrackBuilder_ = &theEventSetup.getData(transientTrackToken_);
+
+  if (cutsFromDB) {
+    hcalCuts = &theEventSetup.getData(hcalCutsToken_);
+  }
 }
 
 void ConvertedPhotonProducer::produce(edm::Event& theEvent, const edm::EventSetup& theEventSetup) {
```
Comment (mmusich) [inline RecoEgamma/EgammaPhotonProducers/src/ConvertedPhotonProducer.cc:202]: for the record this producer could implement a `fillDescription` method.

File: RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc [hunk 13]
```diff
@@ -471,6 +485,12 @@ GEDPhotonProducer::GEDPhotonProducer(const edm::ParameterSet& config, const Cach
   }
 }
 
+void GEDPhotonProducer::beginRun(const edm::Run& run, const edm::EventSetup& eventSetup) {
+  if (cutsFromDB) {
+    hcalCuts = &eventSetup.getData(hcalCutsToken_);
+  }
+}
+
 std::unique_ptr<CacheData> GEDPhotonProducer::initializeGlobalCache(const edm::ParameterSet& config) {
```
Comment (mmusich) [inline RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:494]: for the record this producer could implement a `fillDescription` method.

File: RecoEgamma/EgammaPhotonProducers/src/PhotonProducer.cc [hunk 14]
```diff
@@ -46,6 +49,7 @@ class PhotonProducer : public edm::stream::EDProducer<> {
 public:
   PhotonProducer(const edm::ParameterSet& ps);
```
Comment (mmusich) [inline RecoEgamma/EgammaPhotonProducers/src/PhotonProducer.cc:50]: for the record this producer could implement a `fillDescription` method.

**Review by mmusich — COMMENTED** [review]
just some suggestions for future improvement (not for this PR!)

File: RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc [hunk 15]
```diff
@@ -127,6 +142,7 @@ void EgammaHLTHcalVarProducerFromRecHit::fillDescriptions(edm::ConfigurationDesc
   desc.add<std::vector<double> >("etThresHB", {0, 0, 0, 0});
   desc.add<std::vector<double> >("eThresHE", {0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2});
   desc.add<std::vector<double> >("etThresHE", {0, 0, 0, 0, 0, 0, 0});
+  desc.add<bool>("usePFThresholdsFromDB", true);
```
Comment (Martin-Grunewald) [inline RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc:145]: This makes sure the parameters is inserted if missing from a config, with value "true", so no need to add it explicitly in any cfi file for this plugin.

File: HLTrigger/Configuration/python/HLT_75e33/modules/hltEgammaHoverEL1Seeded_cfi.py [hunk 16]
```diff
@@ -18,6 +18,7 @@
     etThresHB = cms.vdouble( 0.0, 0.0, 0.0, 0.0 ),
     eThresHE = cms.vdouble( 0.1, 0.2, 0.2, 0.2, 0.2, 0.2, 0.2 ),
     etThresHE = cms.vdouble( 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ),
+    usePFThresholdsFromDB = cms.bool(True),
```
[the remark this answers is not in the document]
Comment (Martin-Grunewald) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltEgammaHoverEL1Seeded_cfi.py:21]: Hmm, yes, but now I wonder if this is consistently done, as having a fillDescription and then not adding the parameter in the Phase-2 HLT config will not lead to an error.
Anyway, yes, leave as is here.

[the remark this answers is not in the document]
Comment (mmusich) [reply HLTrigger/Configuration/python/HLT_75e33/modules/hltEgammaHoverEL1Seeded_cfi.py:21]: > but now I wonder if this is consistently done

I share your doubt. Perhaps an item to be brought about at the next TSG coordination meeting in order to clarify a strategy.

**Comment by rappoccio** [comment]
+1

File: RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc [hunk 17]
```diff
@@ -32,12 +32,16 @@ A rho correction can be applied
...
+#include "CondFormats/DataRecord/interface/HcalPFCutsRcd.h"
+#include "CondTools/Hcal/interface/HcalPFCutsHandler.h"
+#include "Geometry/CaloTopology/interface/HcalTopology.h"
 
 class EgammaHLTHcalVarProducerFromRecHit : public edm::global::EDProducer<> {
 public:
   explicit EgammaHLTHcalVarProducerFromRecHit(const edm::ParameterSet &);
 
 public:
+  void beginRun(edm::Run const &, edm::EventSetup const &);
```
Comment (mmusich) [inline RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc:44]: we overlooked that this module is a `edm::global::EDProducer` hence `beginRun` cannot be `override`-n (it's not in the base class). Hence this is never called.

Comment (makortel) [reply RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc:44]: The best fix would be to move the code in `beginRun()` to `produce()` and remove the `hcalCuts` member variable.

Comment (mmusich) [reply RecoEgamma/EgammaHLTProducers/plugins/EgammaHLTHcalVarProducerFromRecHit.cc:44]: https://github.com/cms-sw/cmssw/pull/43819 should fix.

**Pull request merged**